core/include/drivers/versal_nvm.h

```diff
@@ -0,0 +1,290 @@
+	uint8_t prmg_puf_helper_data;
+	uint8_t env_monitor_dis;
+	uint32_t efuse_syn_data[PUF_SYN_DATA_WORDS];
+	uint32_t chash;
+	uint32_t aux;
+	uint8_t pad[56];
+};
+
+struct versal_efuse_puf_user_fuse {
+	uint32_t data_addr[PUF_EFUSES_WORDS];
+	uint8_t env_monitor_dis;
+	uint8_t prgm_puf_fuse;
+	uint32_t start_row;
+	uint32_t num_rows;
+};
+
+struct versal_efuse_write_ops {
+	TEE_Result (*user_data)(uint32_t *p, size_t len, uint32_t first,
+				size_t num);
+	TEE_Result (*boot_env_ctrl)(struct versal_efuse_boot_env_ctrl_bits *p);
+	TEE_Result (*puf_as_user_fuse)(struct versal_efuse_puf_user_fuse *p);
+	TEE_Result (*glitch_cfg)(struct versal_efuse_glitch_cfg_bits *p);
+	TEE_Result (*sec_misc1)(struct versal_efuse_sec_misc1_bits *p);
+	TEE_Result (*misc_ctrl)(struct versal_efuse_misc_ctrl_bits *p);
+	TEE_Result (*offchip_ids)(struct versal_efuse_offchip_ids *p);
+	TEE_Result (*sec_ctrl)(struct versal_efuse_sec_ctrl_bits *p);
+	TEE_Result (*dec_only)(struct versal_efuse_dec_only *p);
+	TEE_Result (*ppk_hash)(struct versal_efuse_ppk_hash *p);
+	TEE_Result (*aes_keys)(struct versal_efuse_aes_keys *p);
+	TEE_Result (*revoke_ppk)(enum versal_nvm_ppk_type type);
+	TEE_Result (*puf)(struct versal_efuse_puf_header *p);
+	TEE_Result (*iv)(struct versal_efuse_ivs *p);
+	TEE_Result (*revoke_id)(uint32_t id);
+};
+
+struct versal_efuse_read_ops {
+	TEE_Result (*offchip_revoke_id)(uint32_t *p, size_t len,
+					enum versal_nvm_offchip_id id);
+	TEE_Result (*revoke_id)(uint32_t *p, size_t len,
+				enum versal_nvm_revocation_id id);
+	TEE_Result (*user_data)(uint32_t *p, size_t len, uint32_t first,
+				size_t num);
+	TEE_Result (*ppk)(uint32_t *p, size_t len,
+			  enum versal_nvm_ppk_type type);
+	TEE_Result (*iv)(uint32_t *p, size_t len, enum versal_nvm_iv_type type);
+	TEE_Result (*boot_env_ctrl)(struct versal_efuse_boot_env_ctrl_bits *p);
+	TEE_Result (*puf_sec_ctrl)(struct versal_efuse_puf_sec_ctrl_bits *p);
+	TEE_Result (*puf_as_user_fuse)(struct versal_efuse_puf_user_fuse *p);
+	TEE_Result (*misc_ctrl)(struct versal_efuse_misc_ctrl_bits *p);
+	TEE_Result (*sec_misc1)(struct versal_efuse_sec_misc1_bits *p);
+	TEE_Result (*sec_ctrl)(struct versal_efuse_sec_ctrl_bits *p);
+	TEE_Result (*puf)(struct versal_efuse_puf_header *p);
+	TEE_Result (*dec_only)(uint32_t *p, size_t len);
+	TEE_Result (*dna)(uint32_t *p, size_t len);
+};
+
+struct versal_bbram_read_ops {
+	TEE_Result (*user_data)(uint32_t *p);
+};
+
+struct versal_bbram_write_ops {
+	TEE_Result (*aes_key)(uint8_t *p, size_t len);
+	TEE_Result (*lock_write_user_data)(void);
+	TEE_Result (*user_data)(uint32_t data);
+	TEE_Result (*zeroize)(void);
+};
+
+struct versal_bbram_ops {
+	struct versal_bbram_write_ops const * const write;
+	struct versal_bbram_read_ops const * const read;
+};
+
+struct versal_efuse_ops {
+	struct versal_efuse_write_ops const * const write;
+	struct versal_efuse_read_ops const * const read;
+};
+
+struct versal_efuse_ops const *versal_get_efuse_ops(void);
```

Comment:
How come you have chosen a function pointer based interface? Do you expect other implementations of this interface?

By the way, given that here's zero documentation of the different functions perhaps you could give a reference to the used interface at least?

Reply:
> How come you have chosen a function pointer based interface? Do you expect other implementations of this interface?

No not really. Was just a matter of readability,  encapsulating the interface in a way that is easy to access and update if needed. Is this an issue

Reply:
> By the way, given that here's zero documentation of the different functions perhaps you could give a reference to the used interface at least?

The information contained in the file is of a similar nature of what you will find on the PLM firmware - I simply removed the redundancy

Take this as an example:
```

/*****************************************************************************/
/**
 * brief	This function sends IPI request to read MiscCtrlBits
 * 		requested by the user
 *
 * param	InstancePtr Pointer to the client instance
 * param	MiscCtrlBits	Address of the output buffer to store the
 * 				MiscCtrlBits eFuses data
 *
 * return	- XST_SUCCESS - If the read is successful
 * 		- XST_FAILURE - If there is a failure
 *
 ******************************************************************************/
int XNvm_EfuseReadMiscCtrlBits(XNvm_ClientInstance *InstancePtr, const u64 MiscCtrlBits)
```

Do you want me to populate OP-TEE with this sort of comments? (not trying to be ironic here...I honestly do not know what you need)

Reply:
> Do you want me to populate OP-TEE with this sort of comments?

No, that doesn't help much. But given that this is an ABI it should be nice to have some kind of documentation to fall back to. Surely there's something you can link to.

Reply:
> > How come you have chosen a function pointer based interface? Do you expect other implementations of this interface?
> 
> No not really. Was just a matter of readability, encapsulating the interface in a way that is easy to access and update if needed. Is this an issue? I can send the HUK driver which uses the interface if it helps with the usability case.

I'm not so keen on function pointer based interfaces when they aren't needed. One of the drawbacks is that linker optimization like garbage collection is prevented with this. Another is that it always raises the question of where are the other implementations, so it's a bit confusing.
We used to have various crypto ops structs but I replaced those in:
commit 3bbd3ce9058c ("core: remove struct crypto_ops")
commit 291e5450a1e9 ("Replace struct acipher_ops with function interface")
commit 33790cc18342 ("Replace struct bignum_ops with function interface")
commit 8875ce46f413 ("Replace struct authenc_ops with function interface")
commit e9eaba5c0a3d ("Replace struct mac_ops with function interface")
commit 82d91db11f01 ("Replace struct cipher_ops with function interface")
commit b887bd8f271b ("Replace struct hash_ops with function interface")

Reply:
oh man, I wish I had known about this restriction. I guess you are asking me to get rid of the pointer based interface then?

> > Do you want me to populate OP-TEE with this sort of comments?
> 
> No, that doesn't help much. But given that this is an ABI it should be nice to have some kind of documentation to fall back to. Surely there's something you can link to.


@nathan-menhorn is there any publicly available document that describes each one of the eFuses?

Reply:
> > > How come you have chosen a function pointer based interface? Do you expect other implementations of this interface?
> > 
> > 
> > No not really. Was just a matter of readability, encapsulating the interface in a way that is easy to access and update if needed. Is this an issue? I can send the HUK driver which uses the interface if it helps with the usability case.
> 
> I'm not so keen on function pointer based interfaces when they aren't needed. One of the drawbacks is that linker optimization like garbage collection is prevented with this. Another is that it always raises the question of where are the other implementations, so it's a bit confusing. We used to have various crypto ops structs but I replaced those in: commit [3bbd3ce](https://github.com/OP-TEE/optee_os/commit/3bbd3ce9058cb0f6c4e7d4725823a3e70029d049) ("core: remove struct crypto_ops") commit [291e545](https://github.com/OP-TEE/optee_os/commit/291e5450a1e9cf4e495d292f40a8ab9b901d9c78) ("Replace struct acipher_ops with function interface") commit [33790cc](https://github.com/OP-TEE/optee_os/commit/33790cc18342231b13f7e6791196bd7c1f41e33f) ("Replace struct bignum_ops with function interface") commit [8875ce4](https://github.com/OP-TEE/optee_os/commit/8875ce46f413b5f5b4e6d73707672d8a3f51f92c) ("Replace struct authenc_ops with function interface") commit [e9eaba5](https://github.com/OP-TEE/optee_os/commit/e9eaba5c0a3d27075b1e995294fd42de65c0b912) ("Replace struct mac_ops with function interface") commit [82d91db](https://github.com/OP-TEE/optee_os/commit/82d91db11f0188df939ce52f6106a39f026de74c) ("Replace struct cipher_ops with function interface") commit [b887bd8](https://github.com/OP-TEE/optee_os/commit/b887bd8f271b056210e964fe7e86dbafe7f6d752) ("Replace struct hash_ops with function interface")

I hear you. I am just the opposite...having an interface to retrieve when needed instead of a direct function call makes the software partitioning cleaner to my eyes. 

Having said I started doing the code with the usual funcion calls but the names of the functions where too long so I switched to that pointer interface...ie: `versal_efuse_read_offchip_revoke_id` or `versal_efuse_write_puf_as_user_fuse`. I think the code is going to be quite ugly hence why I need to make sure this is what you are asking me to do. if you could confirm I'll just rename it all.

Reply:
um I think linker garbage collection can not be used as a silver bullet rule to every interface; in fact I don't think it is that relevant on Versal due to this architecture not being so memory constrained as other platforms (i.e: stm32mp1, imx6ull..). 
Is there any other measurable drawback apart from this one? 

I would like to align with the overall design that you have in mind for OP-TEE as much as possible but at the same time this driver is not part of the OP-TEE generic core. And as I mentioned - and as driver maintainer - in this instance I might value more the clarity in the interface than the few hundreds of bytes that we might save using direct calls.

I suppose if I don't get review comments - and if you dont want to proceed with the review - I will just have to rename the functions and move on. Still have a lot of code pending for Versal in my trees (HUK, ECC, RSA, SHA3...)